MM-42835_Invite People - add email+user invites

julmondragon opened the pull request:
#### Summary
Adding the functionality to invite users by email or invite users from other teams on the server

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-42835

#### Checklist
- [x] Added or updated unit tests (required for all new features)
- [x] Has UI changes
- [x] Includes text changes and localization file updates
- [x] Have tested against the 5 core themes to ensure consistency between them.

#### Device Information
This PR was tested on:
- iPhone 13, iOS 15.2
- Pixel 4, Android 12

#### Screenshots
##### Share link
![Dec-16-2022 12-53-18](https://user-images.githubusercontent.com/79058848/208159970-9908b599-7fe5-4c3e-81f9-5db924a60441.gif)

#### Send email invite
![Dec-16-2022 12-54-58](https://user-images.githubusercontent.com/79058848/208160051-e46ad073-35fa-4fb2-abbc-5d27228e15e7.gif)

#### Invite user to team
![Dec-16-2022 12-55-31](https://user-images.githubusercontent.com/79058848/208160134-200949aa-ad3f-418e-9324-b857d4ae0177.gif)

#### User already in team
![Dec-16-2022 12-57-13](https://user-images.githubusercontent.com/79058848/208160248-fd4d5ee4-93f9-4f10-a2e7-b2b236e87ddc.gif)

#### Handle both sent and not sent invites
![Dec-16-2022 12-56-34](https://user-images.githubusercontent.com/79058848/208160380-79b5a293-744d-4271-8613-71977b910bc6.gif)

#### Release Note
```release-note
- Adding functionality to invite users by email or invite users from other teams on the server
```

mattermod commented:
Building app in separate branch.

mattermod commented:
Successfully building:  
https://circleci.com/gh/mattermost/mattermost-mobile/26959  
https://circleci.com/gh/mattermost/mattermost-mobile/26958

mattermod commented:
Artifact links:  
https://output.circle-artifacts.com/output/job/49b832c8-513b-4d05-825c-2f61ceb6e1ff/artifacts/0/tmp/artifacts/Mattermost_Beta.apk  
https://output.circle-artifacts.com/output/job/35e3c72d-24b9-4b95-9f28-b61b1c9750c0/artifacts/0/tmp/artifacts/Mattermost_Beta.ipa

enahum commented:
@matthewbirtch the input used here does not have a label at the top left, is that expected? If so why? Every other input has it

enahum commented on app/actions/remote/team.ts:
```diff
@@ -98,6 +98,46 @@ export async function addUserToTeam(serverUrl: string, teamId: string, userId: s
+    try {
+        EphemeralStore.startAddingToTeam(teamId);
+
+        const members = await client.addUsersToTeamGracefully(teamId, userIds);
```
If the users are added to the team why are we not adding the relation to the database?

julmondragon replied:
Updated

enahum commented on app/actions/remote/team.ts:
```diff
@@ -98,6 +98,46 @@ export async function addUserToTeam(serverUrl: string, teamId: string, userId: s
+export async function addUsersToTeam(serverUrl: string, teamId: string, userIds: string[]) {
+    let client;
+    try {
+        client = NetworkManager.getClient(serverUrl);
```
I guess this can go in the same try/catch as the rest

julmondragon replied:
Done

enahum commented on app/actions/remote/team.ts:
```diff
@@ -98,6 +98,46 @@ export async function addUserToTeam(serverUrl: string, teamId: string, userId: s
+export async function sendEmailInvitesToTeam(serverUrl: string, teamId: string, emails: string[]) {
+    let client;
+    try {
+        client = NetworkManager.getClient(serverUrl);
```
Same as with the other action

julmondragon replied:
Done

enahum commented on app/actions/remote/team.ts:
```diff
@@ -355,3 +395,21 @@ export async function handleKickFromTeam(serverUrl: string, teamId: string) {
+export async function getTeamMembersByIds(serverUrl: string, teamId: string, userIds: string[]) {
+    let client;
+    try {
+        client = NetworkManager.getClient(serverUrl);
```
This one too

julmondragon replied:
Done

enahum commented on app/actions/remote/team.ts:
```diff
@@ -355,3 +395,21 @@ export async function handleKickFromTeam(serverUrl: string, teamId: string) {
+    }
+
+    try {
+        const members = await client.getTeamMembersByIds(teamId, userIds);
```
Same question about not storing the relation in the database, and more so, if the user is not present, why not adding it?

julmondragon replied:
Updated

enahum commented on app/components/floating_text_input_label/index.tsx:
```diff
@@ -100,7 +100,7 @@ type FloatingTextInputProps = TextInputProps & {
     error?: string;
     errorIcon?: string;
     isKeyboardInput?: boolean;
-    label: string;
```
Not convinced by not having the label, deferred to UX

enahum replied:
UX pointed out that the no label is expected, I this is best to use a TextInput instead of modifying thid component

enahum commented on app/components/illustrations/alert.tsx:
```diff
@@ -0,0 +1,30 @@
+        >
+            <Path
+                d='M4.43715 54.2949C1.46653 54.2949 0.12055 52.1536 1.44612 49.5365L25.4558 2.37353C26.8154 -0.236815 28.9566 -0.236815 30.289 2.37353L54.292 49.5365C55.6515 52.1468 54.292 54.2949 51.3009 54.2949H4.43715Z'
+                fill='#FFBC1F'
```
Are this illustrations theme agnostic?

matthewbirtch replied:
The alert illustration shouldn't need any theme colors applied

enahum commented on app/components/selected_users/selected_user.tsx:
```diff
@@ -27,7 +27,7 @@ type Props = {
      * The user that this component represents.
      */
-    user: UserProfile;
+    user: UserProfile|string;
```
This does not look right to me, if you don't have the profile I rather have a totslly different component

julmondragon replied:
Updated

enahum commented on app/components/user_item/user_item.tsx:
```diff
@@ -80,7 +80,6 @@ const getStyleFromTheme = makeStyleSheetFromTheme((theme: Theme) => {
             color: changeOpacity(theme.centerChannelColor, 0.64),
             fontSize: 15,
             fontFamily: 'OpenSans',
-            flexShrink: 5,
```
What is the impact of this?

enahum replied:
Have you tested this changes in other places where this component is used?

julmondragon replied:
The idea is to shrink this element based on the current elements count, so the space is better distributed. Tested id a well in the autocomplete.

enahum commented on app/components/user_item/user_item.tsx:
```diff
@@ -146,15 +154,15 @@ const UserItem = ({
-                }
+                {Boolean(user) && (
+                    <Text
+                        style={{...style.rowUsername, flexShrink: rowUsernameFlexShrink}}
```
```suggestion
                        style={[style.rowUsername, {flexShrink: rowUsernameFlexShrink}]}
```

And maybe is best to memoize it

julmondragon replied:
Updated

enahum commented on app/screens/home/channel_list/categories_list/header/plus_menu/index.tsx:
```diff
@@ -57,58 +50,18 @@ const PlusMenuList = ({canCreateChannels, canJoinChannels, canInvitePeople, disp
-            },
-        });
+        const title = intl.formatMessage({id: 'invite.title', defaultMessage: 'Invite'});
+        const closeButton = await CompassIcon.getImageSource('close', 24, theme.sidebarHeaderTextColor);
```
There is a sync method you can use

julmondragon replied:
Updated

enahum commented on app/screens/home/channel_list/categories_list/header/plus_menu/index.tsx:
```diff
@@ -57,58 +50,18 @@ const PlusMenuList = ({canCreateChannels, canJoinChannels, canInvitePeople, disp
+        showModal(
+            Screens.INVITE,
+            title,
+            {closeButton},
```
Don't you need to set a closeButtonId as well?

enahum commented on app/screens/invite/index.ts:
```diff
@@ -0,0 +1,41 @@
+            switchMap((t) => of$(t?.inviteId)),
+        ),
+        teammateNameDisplay: observeTeammateNameDisplay(database),
+        isAdmin: observeCurrentUser(database).pipe(
```
Is this only for sysadmins?

julmondragon replied:
No, this is only for a special error handling when no SMTP configured yet.

enahum commented on app/screens/invite/selection.tsx:
```diff
@@ -0,0 +1,512 @@
+                />
+            </Animated.View>
+        </View>
+    );
```
This component is a little bit too big, can you split it in smaller ones or move some functionality out?

julmondragon replied:
Updated

enahum commented on app/screens/navigation.ts:
```diff
@@ -695,7 +695,21 @@ export function setButtons(componentId: string, buttons: NavButtons = {leftButto
 }
 
-export function showOverlay(name: string, passProps = {}, options: Options = {}) {
+export function setTitle(componentId: string, title: string) {
```
Thid is definitely not needed

enahum replied:
Again, this is not needed, in any case this can be called directly in the screen component

julmondragon replied:
Checking this

julmondragon replied:
Updated

enahum reviewed (changes requested):
I did a very high level review, multiple questions and observations

matthewbirtch commented:
> @matthewbirtch the input used here does not have a label at the top left, is that expected? If so why? Every other input has it

Yes @enahum this is intentional and as designed. Because the input field is so close to the title it is a bit redundant to have a label on the input field as well. It's just extra noise that isn't needed in this case.

matthewbirtch reviewed (changes requested):
Nice work on this so far @julmondragon. Just ran through it and have a few thoughts/comments/tweaks to make:

1. Can we auto-focus on the input field when the invite modal opens?
2. There seems to be some positioning issue with the autocomplete when the keyboard is open. The autocomplete popover seems to be hidden in behind the keyboard while the software keyboard is open. Check out this screen recording:

    https://user-images.githubusercontent.com/2040554/208194045-4e1a629e-4e8c-4bbb-ab26-5967993906f4.mov

3. In the autocomplete popover, can we add a bit of vertical padding on the list (8 should do it) of the list to balance the spacing a bit better 
4. While entering a user or email address it seems like there is a significant delay (testing on simulator) before the autocomplete shows results. It doesn't seem to narrow the results at all until I've stopped typing for 1-2 seconds. That feels too long to wait.
5. The autocomplete popover doesn't have a shadow on iOS. I see an shadow added int he code, but it doesn't seem to be rendering. I think there is an overflow clipping this and preventing the shadow from being visible
6. I was able to successfully invite myself even though I'm already on the team. I think it should show as "not sent" with a message that "This person is already on the team" in this case shouldn't it?
    <img width="564" alt="image" src="https://user-images.githubusercontent.com/2040554/208197615-c9535f61-f04b-4a49-96dc-4bac7b3fc3d1.png">
7. If I enter an email address then turn off my wifi and try to send, I get an endless spinner. We should probably timeout here sooner and show the error summary state with an explanation that there was a connectivition issue.
8. On the summary screen, can we add a max width on the main content area? I think we can make it 480.
9. On the summary screen, can we make the bottom margin on the title 24 instead of 16?

I haven't reviewed on Android yet, but can do that once the above are addressed.

julmondragon commented:
@matthewbirtch 
> 1. Can we auto-focus on the input field when the invite modal opens?
  Done
> 2. There seems to be some positioning issue with the autocomplete when the keyboard is open. The autocomplete popover seems to be hidden in behind the keyboard while the software keyboard is open. Check out this screen recording:
  Fixed
> 3. In the autocomplete popover, can we add a bit of vertical padding on the list (8 should do it) of the list to balance the spacing a bit better
  Done
> 4. While entering a user or email address it seems like there is a significant delay (testing on simulator) before the autocomplete shows results. It doesn't seem to narrow the results at all until I've stopped typing for 1-2 seconds. That feels too long to wait.
   Reduced timeout 
> 5. The autocomplete popover doesn't have a shadow on iOS. I see an shadow added int he code, but it doesn't seem to be rendering. I think there is an overflow clipping this and preventing the shadow from being visible
  Fixed
> 6. I was able to successfully invite myself even though I'm already on the team. I think it should show as "not sent" with a message that "This person is already on the team" in this case shouldn't it?
  Updated to not be able to add email of actual user, same as in webapp.
> 7. If I enter an email address then turn off my wifi and try to send, I get an endless spinner. We should probably timeout here sooner and show the error summary state with an explanation that there was a connectivition issue.
  Updated generic service error handling, please check wording for the error message:
  
<img width="435" alt="Screen Shot 2023-01-06 at 5 31 28 PM" src="https://user-images.githubusercontent.com/79058848/211380565-de616f44-cd04-467f-b1f0-596b92f11869.png">

> 8. On the summary screen, can we add a max width on the main content area? I think we can make it 480.
  Done
> 9. On the summary screen, can we make the bottom margin on the title 24 instead of 16?
  Done

mattermost-build commented:
Building app in separate branch.

mattermost-build commented:
Successfully building:  
https://circleci.com/gh/mattermost/mattermost-mobile/27326  
https://circleci.com/gh/mattermost/mattermost-mobile/27325

mattermost-build commented:
Artifact links:  
https://output.circle-artifacts.com/output/job/4e57a3a3-cc56-4455-b27f-92dc2dac085d/artifacts/0/tmp/artifacts/Mattermost_Beta.ipa  
https://output.circle-artifacts.com/output/job/21b620d6-32c3-4276-bd8e-a2ce589954d3/artifacts/0/tmp/artifacts/Mattermost_Beta.apk

matthewbirtch commented:
Thanks @julmondragon. Those changes look good. Just this last one I think:

> > 7\. If I enter an email address then turn off my wifi and try to send, I get an endless spinner. We should probably timeout here sooner and show the error summary state with an explanation that there was a connectivition issue.

> Updated generic service error handling, please check wording for the error message:

Can we update the error screen as follows:

**Invitations could not be sent successfully**
Something went wrong while trying to send invitations. Please check your network connection and try again.
[Try again]

<img src="https://user-images.githubusercontent.com/2040554/211405715-0d8cce5e-5596-4d8b-8e05-b82c70950b2b.png" width="400">


And can we add the 'Back' button in the top nav bar on this screen to go back?

mattermost-build commented:
Building app in separate branch.

julmondragon commented:
@matthewbirtch  
> Can we update the error screen as follows:
> 
> **Invitations could not be sent successfully** Something went wrong while trying to send invitations. Please check your network connection and try again. [Try again]
> 
> And can we add the 'Back' button in the top nav bar on this screen to go back?

Done

<img width="421" alt="Screen Shot 2023-01-15 at 4 58 15 PM" src="https://user-images.githubusercontent.com/79058848/212570501-cf74f3de-d203-4bfb-bb5b-8a329b1e7712.png">

mattermost-build commented:
Successfully building:  
https://circleci.com/gh/mattermost/mattermost-mobile/27473  
https://circleci.com/gh/mattermost/mattermost-mobile/27472

mattermost-build commented:
Artifact links:  
https://output.circle-artifacts.com/output/job/06988b52-5be1-4d6f-aaab-daf1c4894b0b/artifacts/0/tmp/artifacts/Mattermost_Beta.ipa  
https://output.circle-artifacts.com/output/job/d37e2b19-560c-44f2-b9fd-10327a75e3ce/artifacts/0/tmp/artifacts/Mattermost_Beta.apk

julmondragon commented:
@enahum @matthewbirtch @pvev All feedback addressed now.
@stevemudie This should be good to start QA review

matthewbirtch commented:
Thanks @julmondragon. Looks good. I have a question about the 'Try again' button in this new error screen, though. When I tap it, it just takes me back to the invitation screen with nothing filled in. I was more expecting 'Try again' to try sending the invitation(s) again rather than take me back to the previous screen. Can you clarify what action is performed when I tap this button?

stevemudie reviewed (approved):
This looks great @julmondragon!

I tested it on both iOS & Android without any issues.

It's all set from my end but I do have the same question regarding the 'Try again' functionality that Matt raised above.

mattermost-build commented:
Building app in separate branch.

mattermost-build commented:
Successfully building:  
https://circleci.com/gh/mattermost/mattermost-mobile/27490  
https://circleci.com/gh/mattermost/mattermost-mobile/27489

mattermost-build commented:
Artifact links:  
https://output.circle-artifacts.com/output/job/a227f978-4393-43bf-8efb-797bc995e594/artifacts/0/tmp/artifacts/Mattermost_Beta.apk  
https://output.circle-artifacts.com/output/job/26f158d2-4540-40a4-924c-9e1fd255e284/artifacts/0/tmp/artifacts/Mattermost_Beta.ipa

julmondragon commented:
@matthewbirtch @stevemudie  The error screen was updated now to:
- Have two buttons at the bottom: [Go back] [Try again]
- In the top nav bar, having the close button
- [Go back] button will take user to the selection screen preserving the invitations
- [Try again] button will attempt sending the invitations again

![Jan-16-2023 13-09-12](https://user-images.githubusercontent.com/79058848/212761876-689d3825-eed8-44f8-beee-43fdc0f65bee.gif)

matthewbirtch reviewed (approved):
Going to approve this to unblock. I sent you a DM with a couple minor tweaks on tablet possibly, but nothing blocking

stevemudie commented:
The changes look great @julmondragon!

mattermost-build commented:
Building app in separate branch.

mattermost-build commented:
Successfully building:  
https://circleci.com/gh/mattermost/mattermost-mobile/27497  
https://circleci.com/gh/mattermost/mattermost-mobile/27496

mattermost-build commented:
Artifact links:  
https://output.circle-artifacts.com/output/job/c18493d9-c510-47b3-975d-531b62dc0c45/artifacts/0/tmp/artifacts/Mattermost_Beta.ipa  
https://output.circle-artifacts.com/output/job/80dc523d-08ea-46da-bee2-c4b4a0fa9924/artifacts/0/tmp/artifacts/Mattermost_Beta.apk

julmondragon commented:
> Going to approve this to unblock. I sent you a DM with a couple minor tweaks on tablet possibly, but nothing blocking
@matthewbirtch 
cc @stevemudie 

Updated to cover min width issues in tablet when in summary:
<img width="710" alt="Screen Shot 2023-01-16 at 6 58 07 PM" src="https://user-images.githubusercontent.com/79058848/212786025-77c9eb9d-fc36-46ed-a270-15024e626e7d.png">

avinashlng1080 commented on app/actions/remote/team.ts:
```diff
@@ -428,3 +497,50 @@ export async function handleKickFromTeam(serverUrl: string, teamId: string) {
+
+            for (const member of members) {
+                teamMemberships.push(member);
+                member.roles.split(' ').forEach((role) => {
+                    if (!roles[role]) {
+                        roles[role] = true;
+                    }
```
For my understanding, why are we doing this?

julmondragon replied:
Just trying to get unique roles for fetching. Updated for a simpler approach using arrays.

avinashlng1080 reviewed (commented):
LGTM.  Just a couple of questions

jfrerich commented on app/screens/invite/selection.tsx:
```diff
@@ -0,0 +1,392 @@
+        },
+        iconContainer: {
+            width: 40,
+            height: 40,
+        },
+        textContainer: {
+            display: 'flex',
+            flexDirection: 'column',
+        },
+        teamText: {
+            color: theme.centerChannelColor,
+            marginLeft: 12,
+            ...typography('Body', 200, 'SemiBold'),
+        },
+        serverText: {
+            color: changeOpacity(theme.centerChannelColor, 0.72),
+            marginLeft: 12,
+            ...typography('Body', 75, 'Regular'),
+        },
+        shareLink: {
+            display: 'flex',
+            marginLeft: 'auto',
+        },
+        shareLinkButton: {
+            display: 'flex',
+            flexDirection: 'row',
+            alignItems: 'center',
+            height: 40,
+            paddingHorizontal: 20,
+            backgroundColor: changeOpacity(theme.buttonBg, 0.08),
+            borderRadius: 4,
+        },
+        shareLinkText: {
+            color: theme.buttonBg,
+            ...typography('Body', 100, 'SemiBold'),
+            paddingLeft: 7,
+        },
+        shareLinkIcon: {
+            color: theme.buttonBg,
+        },
```
These aren't used in the file.

julmondragon replied:
Remained after splitting the initial component, removed.

jfrerich commented on app/screens/invite/summary_report.tsx:
```diff
@@ -0,0 +1,150 @@
+
+const getStyleSheet = makeStyleSheetFromTheme((theme: Theme) => {
+    return {
+        summaryInvitationsContainer: {
```
all of these contain a `summaryInvitations` prefix and I would suggest removing it from each of the field names.

julmondragon replied:
Removed

jfrerich reviewed (changes requested):
Just a couple of nitpicks and suggestions

pvev reviewed (approved):
thanks! Great work @julmondragon ! :shipit:

mattermost-build commented:
Building app in separate branch.

mattermost-build commented:
Successfully building:  
https://circleci.com/gh/mattermost/mattermost-mobile/27532  
https://circleci.com/gh/mattermost/mattermost-mobile/27531

mattermost-build commented:
Artifact links:  
https://output.circle-artifacts.com/output/job/bea59846-278c-4dc6-b2f2-1223b8c59c06/artifacts/0/tmp/artifacts/Mattermost_Beta.apk  
https://output.circle-artifacts.com/output/job/3055a00f-3902-4622-ac79-8124ca7313c2/artifacts/0/tmp/artifacts/Mattermost_Beta.ipa

julmondragon commented on app/screens/navigation.ts:
```diff
@@ -703,7 +703,7 @@ export function setButtons(componentId: string, buttons: NavButtons = {leftButto
 }
 
-export function showOverlay(name: string, passProps = {}, options: Options = {}) {
+export function showOverlay(name: string, passProps = {}, options = {}) {
```
Remain from initial changes, added back.

julmondragon commented on app/screens/invite/summary.tsx:
```diff
@@ -0,0 +1,295 @@
+                                invites={sent}
+                                selectedIds={selectedIds}
+                                testID='invite.summary_report'
+                            />
```
Makes sense, updated.

julmondragon commented on app/screens/invite/selection_team_bar.tsx:
```diff
@@ -0,0 +1,215 @@
+        Share.open(
+            options,
+        ).catch(() => {
+            // do nothing
```
I don't think there is a handling defined for any failure opening the bottom share, in general in the app.
I can check with UX and try cover it in the following iteration of the feature (WIP).

julmondragon commented on app/screens/invite/invite.tsx:
```diff
@@ -0,0 +1,406 @@
+            return;
+        }
+
+        const {data} = await searchProfiles(serverUrl, searchTerm.toLowerCase(), {allow_inactive: true});
```
Yes, it is my understanding. But here we don't need those, so removing this option.

julmondragon commented on app/screens/invite/invite.tsx:
```diff
@@ -0,0 +1,406 @@
+        setSendError('');
+        setStage(Stage.LOADING);
+
+        setTimeout(() => {
```
Updated.

julmondragon commented on assets/base/i18n/en.json:
```diff
@@ -327,6 +327,28 @@
   "invite_people_to_team.message": "Here’s a link to collaborate and communicate with us on Mattermost.",
   "invite_people_to_team.title": "Join the {team} team",
+  "invite.members.already-member": "This person is already a team member",
+  "invite.members.user-is-guest": "Contact your admin to make this guest a full member",
```
Good catch, updated.

julmondragon commented on app/actions/remote/team.ts:
```diff
@@ -114,6 +114,75 @@ export async function addUserToTeam(serverUrl: string, teamId: string, userId: s
+            const teamMemberships: TeamMembership[] = [];
+            const roles: Record<string, boolean> = {};
+
+            for (const {member} of members) {
```
It is needed to extract the member from the returned `TeamMemberWithError` type. 
Updated user's roles extraction process.

julmondragon replied:
No, member is present. Just contains information about unsuccessful addition to team.

julmondragon commented on app/screens/invite/summary_report.tsx:
```diff
@@ -0,0 +1,150 @@
+                <CompassIcon
+                    name={sent ? 'check-circle' : 'close-circle'}
+                    size={24}
+                    style={{color: sent ? '#3db887' : '#d24b4e'}}
```
Done

julmondragon commented:
@avinashlng1080 @jfrerich @pvev thanks for the feedback, all changes should be covered now.

jfrerich reviewed (approved):
Great work!

enahum commented on app/screens/invite/invite.tsx:
```diff
@@ -0,0 +1,419 @@
+        }
+
+        searchTimeoutId.current = setTimeout(async () => {
+            await searchUsers(text);
```
@larkox this is a good example on why we would need to cancel a previous request.

enahum commented on app/screens/invite/invite.tsx:
```diff
@@ -0,0 +1,419 @@
+            }
+        }
+
+        const currentMemberIds: Record<string, boolean> = {};
```
Would a Set be better here?

julmondragon replied:
Updated

enahum commented on app/screens/invite/invite.tsx:
```diff
@@ -0,0 +1,419 @@
+        };
+
+        setButtons(componentId, buttons);
+    }, [theme, locale, componentId, selectedCount, stage, sendError]);
```
To have this effect executed less times, you can change the dependency to `selectCount > 0` and stage === SELECTION

julmondragon replied:
Updated

enahum commented on app/screens/invite/invite.tsx:
```diff
@@ -0,0 +1,419 @@
+                },
+            },
+        });
+    }, [componentId, locale, theme, stage]);
```
This one too stage === Result

julmondragon replied:
Updated

enahum commented on app/screens/invite/selected_email.tsx:
```diff
@@ -0,0 +1,88 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
```
Can it be done in this PR?

julmondragon replied:
Added reusable `selected_chip`.

enahum commented on app/screens/invite/selection.tsx:
```diff
@@ -0,0 +1,343 @@
+            paddingVertical: 8,
+        },
+        searchListShadow: {
+            shadowColor: '#000',
```
To help with the layout calculation setting a background color always help. If the background color mess things up, leave it out

julmondragon replied:
Good tip, added.

enahum replied:
I don't see it

julmondragon replied:
Added in `searchList`.

enahum replied:
it needs to be on the same style, if it is in a different style it does not get applied in the calculation. So you need to move it to this style and test it, if it does not work correctly, feel free to remove it.

julmondragon replied:
Done.

enahum commented on app/screens/invite/selection.tsx:
```diff
@@ -0,0 +1,343 @@
+        return Math.min(animatedAutocompleteAvailableSpace.value, defaultMaxHeight);
+    }, [animatedAutocompleteAvailableSpace, defaultMaxHeight]);
+
+    const searchListContainerStyle = useMemo(() => {
```
This should be an animated style

julmondragon replied:
Updated

mattermost-build commented:
Building app in separate branch.

mattermost-build commented:
Successfully building:  
https://circleci.com/gh/mattermost/mattermost-mobile/27580  
https://circleci.com/gh/mattermost/mattermost-mobile/27579

mattermost-build commented:
Artifact links:  
https://output.circle-artifacts.com/output/job/a0264ada-4145-44b8-864e-4779c18d3a72/artifacts/0/tmp/artifacts/Mattermost_Beta.apk  
https://output.circle-artifacts.com/output/job/2e4e32d0-d2d2-42b8-82e2-641926859a65/artifacts/0/tmp/artifacts/Mattermost_Beta.ipa

julmondragon commented on app/actions/remote/team.ts:
```diff
@@ -114,6 +114,75 @@ export async function addUserToTeam(serverUrl: string, teamId: string, userId: s
+            setTeamLoading(serverUrl, true);
+
+            const teamMemberships: TeamMembership[] = [];
+            const roles: Record<string, boolean> = {};
```
Updated

julmondragon commented on app/actions/remote/team.ts:
```diff
@@ -114,6 +114,75 @@ export async function addUserToTeam(serverUrl: string, teamId: string, userId: s
+
+            fetchRolesIfNeeded(serverUrl, Object.getOwnPropertyNames(roles));
+
+            const operator = DatabaseManager.serverDatabases[serverUrl]?.operator;
```
Updated

julmondragon commented on app/actions/remote/team.ts:
```diff
@@ -114,6 +114,75 @@ export async function addUserToTeam(serverUrl: string, teamId: string, userId: s
+            const operator = DatabaseManager.serverDatabases[serverUrl]?.operator;
+
+            if (operator) {
+                const team = await client.getTeam(teamId);
```
Removed fetch for team.

julmondragon commented on app/actions/remote/team.ts:
```diff
@@ -114,6 +114,75 @@ export async function addUserToTeam(serverUrl: string, teamId: string, userId: s
+        EphemeralStore.finishAddingToTeam(teamId);
+        return {members};
+    } catch (error) {
+        if (client) {
```
It was to be sure `startAddingToTeam` was set. Updated to use  `isAddingToTeam`.

julmondragon commented on app/actions/remote/team.ts:
```diff
@@ -114,6 +114,75 @@ export async function addUserToTeam(serverUrl: string, teamId: string, userId: s
+            EphemeralStore.finishAddingToTeam(teamId);
+        }
+
+        forceLogoutIfNecessary(serverUrl, error as ClientError);
```
At this phase I don't think so. Will check in general in the next phase.

julmondragon commented on app/actions/remote/team.ts:
```diff
@@ -428,3 +497,44 @@ export async function handleKickFromTeam(serverUrl: string, teamId: string) {
+
+            fetchRolesIfNeeded(serverUrl, Array.from(new Set(roles)));
+
+            const operator = DatabaseManager.serverDatabases[serverUrl]?.operator;
```
Updated

julmondragon commented on app/actions/remote/team.ts:
```diff
@@ -114,6 +114,75 @@ export async function addUserToTeam(serverUrl: string, teamId: string, userId: s
+        const members = await client.addUsersToTeamGracefully(teamId, userIds);
+
+        if (!fetchOnly) {
+            setTeamLoading(serverUrl, true);
```
Removed Team fetch and related.

julmondragon commented on app/actions/remote/team.ts:
```diff
@@ -428,3 +497,44 @@ export async function handleKickFromTeam(serverUrl: string, teamId: string) {
+        const members = await client.getTeamMembersByIds(teamId, userIds);
+
+        if (!fetchOnly) {
+            setTeamLoading(serverUrl, true);
```
Removed Team fetch and related.

julmondragon commented on app/actions/remote/team.ts:
```diff
@@ -428,3 +497,44 @@ export async function handleKickFromTeam(serverUrl: string, teamId: string) {
+            const operator = DatabaseManager.serverDatabases[serverUrl]?.operator;
+
+            if (operator) {
+                const team = await client.getTeam(teamId);
```
Removed Team fetch and related.

julmondragon commented on app/client/rest/teams.ts:
```diff
@@ -130,6 +140,27 @@ const ClientTeams = (superclass: any) => class extends superclass {
+        const members: Array<{team_id: string; user_id: string}> = [];
+        userIds.forEach((id) => members.push({team_id: teamId, user_id: id}));
+
+        return this.doFetch(
```
- `/ids` v3.8.0
- `/batch?graceful=` v5.22.0
- `/invite/email?graceful=` v5.22.0

julmondragon commented on app/components/user_item/user_item.tsx:
```diff
@@ -148,15 +160,15 @@ const UserItem = ({
-                }
+                {Boolean(user) && (
+                    <Text
+                        style={usernameTextStyle}
```
No change, just to improve space use. Checked other places using it.

julmondragon commented on app/screens/home/channel_list/categories_list/header/header.tsx:
```diff
@@ -97,7 +97,6 @@ const ChannelListHeader = ({
     canJoinChannels,
     canInvitePeople,
     displayName,
-    inviteId,
```
Done

julmondragon commented:
@enahum Thanks for the feedback, all should be addressed now.

enahum commented on app/actions/remote/team.ts:
```diff
@@ -114,6 +114,59 @@ export async function addUserToTeam(serverUrl: string, teamId: string, userId: s
+}
+
+export async function sendEmailInvitesToTeam(serverUrl: string, teamId: string, emails: string[]) {
+    let client;
+
+    try {
+        client = NetworkManager.getClient(serverUrl);
```
```suggestion
    try {
        const client = NetworkManager.getClient(serverUrl);
```

julmondragon replied:
Done

enahum commented on app/actions/remote/team.ts:
```diff
@@ -428,3 +481,32 @@ export async function handleKickFromTeam(serverUrl: string, teamId: string) {
 }
+
+export async function getTeamMembersByIds(serverUrl: string, teamId: string, userIds: string[], fetchOnly?: boolean) {
+    let client;
+
+    try {
+        client = NetworkManager.getClient(serverUrl);
```
```suggestion
    try {
        const client = NetworkManager.getClient(serverUrl);
```

julmondragon replied:
Done

enahum commented on app/actions/remote/team.ts:
```diff
@@ -428,3 +481,32 @@ export async function handleKickFromTeam(serverUrl: string, teamId: string) {
+
+            fetchRolesIfNeeded(serverUrl, Array.from(new Set(roles)));
+
+            if (operator) {
+                await operator.handleTeamMemberships({teamMemberships: members, prepareRecordsOnly: true});
+            }
```
`operator` is always defined here.

```suggestion
            await operator.handleTeamMemberships({teamMemberships: members, prepareRecordsOnly: true});
```

julmondragon replied:
Done

enahum reviewed (changes requested):
Great work.. almost there...

mattermost-build commented:
Building app in separate branch.

mattermost-build commented:
Successfully building:  
https://circleci.com/gh/mattermost/mattermost-mobile/27598  
https://circleci.com/gh/mattermost/mattermost-mobile/27597

enahum reviewed (approved):
Awesome

mattermost-build commented:
Artifact links:  
https://output.circle-artifacts.com/output/job/d2088ae7-3623-4bb7-bac9-ba84191e5ff9/artifacts/0/tmp/artifacts/Mattermost_Beta.ipa  
https://output.circle-artifacts.com/output/job/2ad8a455-3ead-4afa-9d70-184cb6ca3f01/artifacts/0/tmp/artifacts/Mattermost_Beta.apk

esethna commented:
@jfrerich are we reintroducing the team drop down now that the view members, invite people options are in the product?

matthewbirtch commented:
> @jfrerich are we reintroducing the team drop down now that the view members, invite people options are in the product?

I don't think view team members is available, just view channel members @esethna

jfrerich commented:
I cannot attest to the view team members, but we do will have view channel members.
